File: 208.实现-trie-前缀树.go
/*
 * @lc app=leetcode.cn id=208 lang=golang
 *
 * [208] 实现 Trie (前缀树)
 */

// @lc code=start
package main
import "strings"


type Trie struct {
	m map[string]bool
}


/** Initialize your data structure here. */
func Constructor() Trie {
	m := map[string]bool{}
	return Trie{m:m}
}


/** Inserts a word into the trie. */
func (this *Trie) Insert(word string)  {
	this.m[word] = true
}


/** Returns if the word is in the trie. */
func (this *Trie) Search(word string) bool {
	if _,ok := this.m[word] ; ok {
		return true
	}
	return false
}


/** Returns if there is any word in the trie that starts with the given prefix. */
func (this *Trie) StartsWith(prefix string) bool {
	for key,_ := range this.m {
		if strings.HasPrefix(key,prefix) {
			return true
		}
	}
	return false
}


/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */
// @lc code=end

File: 208.实现-trie-前缀树.py
#
# @lc app=leetcode.cn id=208 lang=python3
#
# [208] 实现 Trie (前缀树)
#

# @lc code=start
class Node:
    def __init__(self):
        self.child = [None] * 26
        self.is_word = False

class Trie:

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        cur = self.root
        for b in word:
            i = ord(b) - ord('a')
            if not cur.child[i]:
                cur.child[i] = Node()
            cur = cur.child[i]
        cur.is_word = True

    def search(self, word: str) -> bool:
        cur = self.root
        for b in word:
            i = ord(b) - ord('a')
            if not cur.child[i]:
                return False
            cur = cur.child[i]
        return cur.is_word

    def startsWith(self, prefix: str) -> bool:
        cur = self.root
        for b in prefix:
            i = ord(b) - ord('a')
            if not cur.child[i]:
                return False
            cur = cur.child[i]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
# @lc code=end

